micronota/parsers/gff.py

```diff
@@ -0,0 +1,239 @@
+documentation.
+
+Derives_from
+Used to disambiguate the relationship between one feature and another when the
+relationship is a temporal one rather than a purely structural "part of" one.
+This is needed for polycistronic genes.
+
+Note
+A free text note.
+
+Dbxref
+A database cross reference. See the GFF3 specification for more information.
+
+Ontology_term
+A cross reference to an ontology term.
+
+Multiple attributes of the same type are indicated by separating
+the values with the comma ","
+
+
+Format Support
+--------------
+**Has Sniffer: Yes**
+
++------+------+---------------------------------------------------------------+
+|Reader|Writer|                          Object Class                         |
++======+======+===============================================================+
+|Yes   |No    |:mod:`skbio.metadata.IntervalMetadata` objects                 |
++------+------+---------------------------------------------------------------+
+|Yes   |No    |generator of :mod:`skbio.metadata.IntervalMetadata` objects    |
++------+------+---------------------------------------------------------------+
+
+
+Reference
+---------
+.. [#] http://gmod.org/wiki/GFF3
+'''
+
+from skbio.util._misc import merge_dicts
+from skbio.io import create_format, FileFormatError
+from skbio.metadata import IntervalMetadata, Feature
+from skbio.io.format._base import (
+    _line_generator, _get_nth_sequence, _too_many_blanks)
+
+
+class GFFFormatError(FileFormatError):
+    pass
+
+gff = create_format('gff')
+
+# Annotation headers
+_ANNOTATION_HEADERS = [
+          'SEQID',
+          'SOURCE',
+          'TYPE',
+          'SCORE',
+          'STRAND',
+          'PHASE',
+          'ATTR'
+          ]
+
+# Attribute headers
+_ATTR_HEADERS = [
+        'ID',
+        'Name',
+        'Alias',
+        'Parent',
+        'Target',
+        'Gap',
+        'Derives_from',
+        'Note',
+        'Dbxref',
+        'Ontology_term'
+        ]
+
+
+@gff.sniffer()
+def _gff_sniffer(fh):
+    # check the 1st real line is a valid ID line
+    if _too_many_blanks(fh, 5):
+        return False, {}
+
+    try:
+        line = next(_line_generator(fh, skip_blanks=True, strip=False))
+    except StopIteration:
+        return False, {}
+
+    try:
+        assert line.startswith('##gff-version 3')
+    except GFFFormatError:
+        return False, {}
+    return True, {}
+
+
+def _is_float(input):
+    try:
+        float(input)
+    except ValueError:
+        return False
+    return True
+
+
+def parse_attr(s):
+    _field = s.split(';')
+    _attr = ()
+    for f in _field:
+        try:
+            _type, _val = f.split('=')
+        except ValueError:
+            _type, _val = 0, 0
+
+        _attr = _attr + (_type, _val)
+    return _attr
+
+
+def parse_required(s):
+    if s.isdigit():
+        return int(s)
+    elif _is_float(s):
+        return float(s)
+    else:
+        return s
+
+
+def IntervalMetadata_construct(_attr, _interval):
```

Comment:
and try to use the form of `verb + norm (optional)` as func name. and lowercase. minor thing, but usually make things clearer